src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/LockFactory.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Razor;
+
+internal class LockFactory
+{
+    // Specify recursion is supported, since an item with an upgradeable lock can still
+    // get another read lock on the same thread
+    protected readonly ReaderWriterLockSlim _lock = new(LockRecursionPolicy.SupportsRecursion);
```

Comment:
Are there descendants of this class? If not, should this field be private?

Reply:
When I was toying with design I had two descendants, but found it unnecessary and more confusing. Changed to private